trendmicro_deepsecurity/icon_trendmicro_deepsecurity/actions/search_rules/action.py

```diff
@@ -16,57 +17,131 @@ def __init__(self):
                 input=SearchRulesInput(),
                 output=SearchRulesOutput())
 
-    def run(self, params={}):
+
+    def search_rule_by_cve(self, cve: str) -> Tuple[set, set, set]:
         '''
-        Searches for IPS rules by CVE number in Deep Security
+        Search IPS rules for one CVE
         '''
 
-        # Get parameters
-        self.vulnerabilities = params.get(Input.VULNERABILITIES)
-
         matched_cves = set()
         missed_cves = set()
         ips_rules = set()
 
-        for cve in self.vulnerabilities:
+        # Prepare Search Criteria
+        data = {"maxItems": 100,
+                "searchCriteria": [{"fieldName": "CVE",
+                                    "stringWildcards": True,
+                                    "stringValue": f"%{cve}%"}]}
 
-            # Prepare request
-            url = f"{self.connection.dsm_url}/api/intrusionpreventionrules/search"
+        # Prepare request
+        url = f"{self.connection.dsm_url}/api/intrusionpreventionrules/search"
+
+        # Search for IPS rules
+        response = self.connection.session.post(url,
+                                                data=json.dumps(data),
+                                                verify=self.connection.dsm_verify_ssl)
 
-            data = {"maxItems": 100,
-                    "searchCriteria": [{"fieldName": "CVE",
-                                        "stringWildcards": True,
-                                        "stringValue": f"%{cve}%"}]}
+        # Check response errors
+        checkResponse(response)
+
+        # Try to convert the response data to JSON
+        response_data = tryJSON(response)
+
+        # Check if matching IPS rules were found
+        if response_data["intrusionPreventionRules"]:
+            hits = len(response_data["intrusionPreventionRules"])
+            for rule in response_data["intrusionPreventionRules"]:
+                self.logger.info(f"{cve} -> {rule['ID']}: {rule['name']}")
+                ips_rules.add(rule["ID"])
+                matched_cves.add(cve)
+        else:
+            self.logger.info(f"{cve}: No rules found!")
+            missed_cves.add(cve)
+
+        return ips_rules, matched_cves, missed_cves
+
+
+    def collect_all_ips_rules(self) -> list:
+        '''
+        Receive all IPS rules from Deep Security
+        '''
 
-            # Search for IPS rules
-            response = self.connection.session.post(url,
-                                                    data=json.dumps(data),
-                                                    verify=self.connection.dsm_verify_ssl)
+        ips_rules = []
+        ips_id = 0
+        step = 5000 # <= 5000
 
-            self.logger.info(f"CVE: {cve}")
-            self.logger.info(f"url: {response.url}")
-            self.logger.info(f"status: {response.status_code}")
-            self.logger.info(f"reason: {response.reason}")
+        while True:         
```

Comment:
This loop isn't clear to me. 

I see the break below, but I'm not sure what you're incrementing or changing to make that break ever be hit. 

It looks like you're making the same request over and over and pulling the last ID out of it. 

Suggestion: Don't use While True unless you absolutely have to, make the "True" part an actual If statement for the break condition.

Reply:
Instead of doing one search API request for every CVE, I receive a list of all rules from Deep Security. 
As the API only allows me to receive 5000 rules with one query I request the rules in batches of 5000 until there are no rules left so response_data["intrusionPreventionRules"] will be empty and break the loop.

Reply:
And as I don't know how many rules there are / or will be in the future I need to use the while true. 
But I'm open to suggestions on how to solve it in a better way.

Reply:
I tries to use
`while response_data["intrusionPreventionRules"]:`
but then I get a "local variable 'response_data' referenced before assignment" error. 

I revert back to the `while True`, but I'm open to other solutions.

Reply:
I think `while True` is fine in this instance... is there anything else you can add to the else clause to ensure a loop break?

Reply:
The loop will break as soon as the request fails or there is no data returned.